### PR TITLE
47305/47306 Health check RNW database and FSS

### PR DESCRIPTION
**Pull request by vipin14904**

### CODE REVIEW

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        public void Setup()
+        {
+            _fakeLogger = A.Fake<ILogger<FileShareServiceHealthCheck>>();
+            _fakeFileShareServiceHealthClient = A.Fake<IFileShareServiceHealthClient>();
+
+            _fileShareServiceHealthCheck = new FileShareServiceHealthCheck(_fakeFileShareServiceHealthClient, _fakeLogger);
+        }
+
+        [Test]
+        public async Task WhenFssConnectionIsHealthy_ThenReturnsHealthy()
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs:28]: Refer Fss as FSS please.
```suggestion
        public async Task WhenFSSConnectionIsHealthy_ThenReturnsHealthy()
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+            HealthCheckResult response = await _fileShareServiceHealthCheck.CheckHealthAsync(new HealthCheckContext());
+
+            Assert.AreEqual(HealthStatus.Healthy, response.Status);
+        }
+
+        [Test]
+        public async Task WhenFssConnectionIsUnHealthy_ThenReturnsUnHealthy()
+        {
+            A.CallTo(() => _fakeFileShareServiceHealthClient.CheckHealthAsync()).Returns(new HealthCheckResult(HealthStatus.Unhealthy, "File share service is unhealthy", new Exception("File share service is unhealthy")));
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs:40]:
```suggestion
            A.CallTo(() => _fakeFileShareServiceHealthClient.CheckHealthAsync()).Returns(new HealthCheckResult(HealthStatus.Unhealthy, "File Share Service is unhealthy", new Exception("File Share Service is unhealthy")));
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            HealthCheckResult healthCheckResult = await _fileShareServiceHealthClient.CheckHealthAsync();
+            if (healthCheckResult.Status == HealthStatus.Healthy)
+            {
+                _logger.LogDebug(EventIds.FileShareServiceIsHealthy.ToEventId(), "File share service is healthy");
+            }
+            else
+            {
+                _logger.LogError(EventIds.FileShareServiceIsUnHealthy.ToEventId(), healthCheckResult.Exception, "File share service is unhealthy responded with error {Message}", healthCheckResult.Exception.Message);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs:27]:
```suggestion
                _logger.LogError(EventIds.FileShareServiceIsUnHealthy.ToEventId(), healthCheckResult.Exception, "File Share Service is unhealthy responded with error {Message}", healthCheckResult.Exception.Message);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        private readonly IAuthFssTokenProvider _authFssTokenProvider;
+
+        public FileShareServiceHealthClient(IHttpClientFactory httpClientFactory, IOptions<FileShareServiceConfiguration> fileShareServiceConfig, IAuthFssTokenProvider authFssTokenProvider)
+        {
+            _httpClientFactory = httpClientFactory;
+            _fileShareServiceConfig = fileShareServiceConfig;
+            _authFssTokenProvider = authFssTokenProvider;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync()
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs:25]:
```suggestion
        public async Task<HealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken)
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            try
+            {
+                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(Guid.NewGuid().ToString());
+                FileShareApiClient fileShareApiClient = new(_httpClientFactory, _fileShareServiceConfig.Value.BaseUrl, accessToken);
+
+                IResult<BatchSearchResponse> result = await FSSSearchAsync(fileShareApiClient);
+
+                if (result.IsSuccess)
+                {
+                    return HealthCheckResult.Healthy("File share service is healthy");
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs:36]: Please change  File share service to File Share Service at all places.
```suggestion
                    return HealthCheckResult.Healthy("File Share Service is healthy");
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    return HealthCheckResult.Unhealthy("File share service is unhealthy", new Exception("Batch search response is not success, Please check configuration."));
+                }
+            }
+            catch (Exception ex)
+            {
+                return HealthCheckResult.Unhealthy("File share service is unhealthy", new Exception(ex.Message));
+            }
+        }
+
+        private async Task<IResult<BatchSearchResponse>> FSSSearchAsync(IFileShareApiClient fileShareApiClient)
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs:49]:
```suggestion
        private async Task<IResult<BatchSearchResponse>> FSSSearchAsync(IFileShareApiClient fileShareApiClient, CancellationToken cancellationToken)
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            string searchQuery = $"BusinessUnit eq 'invalid'";
+            IResult<BatchSearchResponse> result = await fileShareApiClient.Search(searchQuery, _fileShareServiceConfig.Value.PageSize, _fileShareServiceConfig.Value.Start, CancellationToken.None);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs:52]:
```suggestion
            IResult<BatchSearchResponse> result = await fileShareApiClient.Search(searchQuery, _fileShareServiceConfig.Value.PageSize, _fileShareServiceConfig.Value.Start, cancellationToken);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+        public RNWDatabaseHealthCheck(IRNWDatabaseHealthClient rnwDatabaseHealthClient, ILogger<RNWDatabaseHealthCheck> logger)
+        {
+            _rnwDatabaseHealthClient = rnwDatabaseHealthClient;
+            _logger = logger;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            HealthCheckResult healthCheckResult = await _rnwDatabaseHealthClient.CheckHealthAsync();
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs:20]: Pass cancellationToken in all subsequent calls.
```suggestion
            HealthCheckResult healthCheckResult = await _rnwDatabaseHealthClient.CheckHealthAsync(cancellationToken);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+            if (healthCheckResult.Status == HealthStatus.Healthy)
+            {
+                _logger.LogDebug(EventIds.RNWDatabaseIsHealthy.ToEventId(), "Radio navigational Warning database is healthy");
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs:23]: Please change Radio navigational Warning to Radio Navigational Warnings database at all places.
```suggestion
                _logger.LogDebug(EventIds.RNWDatabaseIsHealthy.ToEventId(), "Radio Navigational Warnings database is healthy");
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            }
+            else
+            {
+                _logger.LogError(EventIds.RNWDatabaseIsUnHealthy.ToEventId(), healthCheckResult.Exception, "Radio mavigational warning database is unhealthy responded with error {Message}", healthCheckResult.Exception.Message);
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs:27]: Typo.
```suggestion
                _logger.LogError(EventIds.RNWDatabaseIsUnHealthy.ToEventId(), healthCheckResult.Exception, "Radio Navigational Warnings database is unhealthy responded with error {Message}", healthCheckResult.Exception.Message);
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            _context = context;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync()
+        {
+            try
+            {
+                await _context.WarningType.ToListAsync();
+                return HealthCheckResult.Healthy("Radio navigational warning database is healthy");
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs:20]:
```suggestion
                return HealthCheckResult.Healthy("Radio Navigational Warnings database is healthy");
```

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+            }
+            catch (Exception ex)
+            {
+                return HealthCheckResult.Unhealthy("Radio navigational warning database is unhealthy", new Exception(ex.Message));
```
Comment (usha13529) [inline UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs:24]:
```suggestion
                return HealthCheckResult.Unhealthy("Radio Navigational Warnings database is unhealthy", new Exception(ex.Message));
```

**Review by usha13529 — CHANGES_REQUESTED** [review]
Few comments to fix please.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+        public FileShareServiceHealthCheck(IFileShareServiceHealthClient fileShareServiceHealthClient, ILogger<FileShareServiceHealthCheck> logger)
+        {
+            _fileShareServiceHealthClient = fileShareServiceHealthClient;
+            _logger = logger;
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
+        {
+            HealthCheckResult healthCheckResult = await _fileShareServiceHealthClient.CheckHealthAsync();
```
[the remark this answers is not in the document]
Comment (barrie-cooper) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs:20]: @usha13529, should this not be:
```suggestion
            HealthCheckResult healthCheckResult = await _fileShareServiceHealthClient.CheckHealthAsync(cancellationToken);
```

[the remark this answers is not in the document]
Comment (usha13529) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs:20]: Yes @barrie-cooper, you're right. This should be cancellationToken passed from inline method.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs (hunk 1 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs]: changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        {
+            A.CallTo(() => _fakeFileShareServiceHealthClient.CheckHealthAsync()).Returns(new HealthCheckResult(HealthStatus.Healthy));
+
+            HealthCheckResult response = await _fileShareServiceHealthCheck.CheckHealthAsync(new HealthCheckContext());
+
+            Assert.AreEqual(HealthStatus.Healthy, response.Status);
+        }
+
+        [Test]
+        public async Task WhenFssConnectionIsUnHealthy_ThenReturnsUnHealthy()
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/FileShareServiceHealthCheckTest.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/RNWDatabaseHealthCheckTest.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+            HealthCheckResult response = await _rnwDatabaseHealthCheck.CheckHealthAsync(new HealthCheckContext());
+
+            Assert.AreEqual(HealthStatus.Healthy, response.Status);
+        }
+
+        [Test]
+        public async Task WhenRNWDatabaseIsUnhealthy_ThenReturnsUnhealthy()
+        {
+            A.CallTo(() => _fakeRnwDatabaseHealthClient.CheckHealthAsync()).Returns(new HealthCheckResult(HealthStatus.Unhealthy, "Radio navigational Warning database is unhealthy", new Exception("Radio navigational Warning database is unhealthy")));
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common.UnitTests/HealthCheck/RNWDatabaseHealthCheckTest.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs (hunk 3 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs (hunk 4 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+            if (healthCheckResult.Status == HealthStatus.Healthy)
+            {
+                _logger.LogDebug(EventIds.FileShareServiceIsHealthy.ToEventId(), "File share service is healthy");
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthCheck.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        public async Task<HealthCheckResult> CheckHealthAsync()
+        {
+            try
+            {
+                string accessToken = await _authFssTokenProvider.GenerateADAccessToken(Guid.NewGuid().ToString());
+                FileShareApiClient fileShareApiClient = new(_httpClientFactory, _fileShareServiceConfig.Value.BaseUrl, accessToken);
+
+                IResult<BatchSearchResponse> result = await FSSSearchAsync(fileShareApiClient);
```
[the remark this answers is not in the document]
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs (hunk 5 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs (hunk 6 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/FileShareServiceHealthClient.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs (hunk 8 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs (hunk 10 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthCheck.cs]: Changes done.

File: UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs (hunk 12 above)
Comment (Ashish15251) [reply UKHO.MaritimeSafetyInformation.Web/UKHO.MaritimeSafetyInformation.Common/HealthCheck/RNWDatabaseHealthClient.cs]: Changes done.

**Review by usha13529 — APPROVED** [review]
Thanks @Ashish15251 for fixing review comments. Looks good to me.

**Pull request merged**